eventlog/eventlog.go

```diff
@@ -0,0 +1,56 @@
+// Copyright 2020 GRAIL, Inc. All rights reserved.
+// Use of this source code is governed by the Apache 2.0
+// license that can be found in the LICENSE file.
+
+// Package eventlog provides logging of semi-structured events, particularly in
+// service of downstream analysis, e.g. when machines are started, when a user
+// issues a command, when failures happen.
+//
+// For example, you can log events to CloudWatch Logs:
+//
+//  sess := session.NewSession()
+//  cw := cloudwatchlogs.New(sess)
+//  e := NewCloudWatchEventer(cw, "myLogGroup", "myLogStream")
+//  e.Event("rpcRetry", "org", "GRAIL", "retry", 0, "maxRetry", 10)
+//  e.Event("machineStopped", "addr", "192.168.1.1", "duration", 3600.0, "startTime": 1584140534)
+//
+// These events can now be analyzed and monitored using CloudWatch Logs tooling.
+package eventlog
+
+import (
+	"github.com/grailbio/base/config"
+)
+
+func init() {
+	config.Register("eventer/nop", func(constr *config.Constructor) {
+		constr.Doc = "eventer/nop configures a no-op event logger"
+		constr.New = func() (interface{}, error) {
+			return Nop{}, nil
+		}
+	})
+	// We do the most conservative thing by default, making event logging a
+	// no-op.
+	config.Default("eventer", "eventer/nop")
+}
+
+// Eventer is called to log events.
+type Eventer interface {
```

Comment:
Do we expect this functionally to be used only by bigslice internal code or by users of bigslice? Do we expect to mainly use this base API, or will we want any additional layers on top. For example, we could have different functions that add some common useful k:v pairs.

Reply:
My plan is for this to be just used by bigmachine and bigslice for now to give us time to work out possible operational issues, etc. If we find that the mechanism is very useful, I'd invite others to start using it, though with their own `Eventer` instances. At that point, we might want to expand the API, but I'd wait until then.

As always, my plans are up for discussion, so let me know if you have different ideas.